File: repos/rappterbook-agent/typescript/src/security/approvals.ts
/**
 * Exec Approvals System
 * Manages tool/command execution policies
 */

export type ApprovalPolicy = 'deny' | 'allowlist' | 'full';

export interface ApprovalRule {
  id: string;
  name: string;
  description?: string;
  policy: ApprovalPolicy;

  // Scope
  tools?: string[];
  channels?: string[];
  senders?: string[];
  agents?: string[];

  // Allowlist (only for 'allowlist' policy)
  allowedTools?: string[];
  allowedCommands?: string[];
  allowedPatterns?: string[];

  // Blocklist (applies to all policies)
  blockedTools?: string[];
  blockedCommands?: string[];
  blockedPatterns?: string[];

  // Approval settings
  requireApproval?: boolean;
  approvalTimeout?: number;

  priority: number;
  enabled: boolean;
}

export interface ApprovalRequest {
  id: string;
  ruleId?: string;
  toolName: string;
  toolArgs: Record<string, unknown>;
  sessionId?: string;
  senderId?: string;
  channelId?: string;
  agentId?: string;
  status: 'pending' | 'approved' | 'rejected' | 'expired';
  createdAt: string;
  expiresAt: string;
  resolvedAt?: string;
  resolvedBy?: string;
  reason?: string;
}

export interface ApprovalContext {
  toolName: string;
  toolArgs: Record<string, unknown>;
  sessionId?: string;
  senderId?: string;
  channelId?: string;
  agentId?: string;
}

export interface ApprovalResult {
  allowed: boolean;
  requiresApproval: boolean;
  rule?: ApprovalRule;
  requestId?: string;
  reason?: string;
}

const DEFAULT_APPROVAL_TIMEOUT = 300000; // 5 minutes

export class ApprovalManager {
  private rules = new Map<string, ApprovalRule>();
  private requests = new Map<string, ApprovalRequest>();
  private defaultPolicy: ApprovalPolicy = 'allowlist';
  private approvalCallbacks = new Map<
    string,
    { resolve: (approved: boolean) => void; timeout: NodeJS.Timeout }
  >();

  /**
   * Set default policy
   */
  setDefaultPolicy(policy: ApprovalPolicy): void {
    this.defaultPolicy = policy;
  }

  /**
   * Add an approval rule
   */
  addRule(rule: ApprovalRule): void {
    this.rules.set(rule.id, rule);
  }

  /**
   * Remove an approval rule
   */
  removeRule(ruleId: string): boolean {
    return this.rules.delete(ruleId);
  }

  /**
   * Get a rule
   */
  getRule(ruleId: string): ApprovalRule | undefined {
    return this.rules.get(ruleId);
  }

  /**
   * Get all rules
   */
  getRules(): ApprovalRule[] {
    return Array.from(this.rules.values()).sort((a, b) => b.priority - a.priority);
  }

  /**
   * Enable/disable a rule
   */
  setRuleEnabled(ruleId: string, enabled: boolean): boolean {
    const rule = this.rules.get(ruleId);
    if (!rule) return false;

    rule.enabled = enabled;
    return true;
  }

  /**
   * Check if a tool call is allowed
   */
  checkApproval(context: ApprovalContext): ApprovalResult {
    // Find matching rule
    const rule = this.findMatchingRule(context);

    // If no rule, use default policy
    const policy = rule?.policy ?? this.defaultPolicy;

    // Check blocked list first (applies to all policies)
    if (this.isBlocked(context, rule)) {
      return {
        allowed: false,
        requiresApproval: false,
        rule,
        reason: 'Tool is blocked by policy',
      };
    }

    switch (policy) {
      case 'deny':
        return {
          allowed: false,
          requiresApproval: false,
          rule,
          reason: 'Tool execution denied by policy',
        };

      case 'allowlist':
        if (this.isAllowed(context, rule)) {
          // Check if still requires explicit approval
          if (rule?.requireApproval) {
            return {
              allowed: false,
              requiresApproval: true,
              rule,
              reason: 'Tool requires explicit approval',
            };
          }
          return { allowed: true, requiresApproval: false, rule };
        }
        return {
          allowed: false,
          requiresApproval: true,
          rule,
          reason: 'Tool not in allowlist',
        };

      case 'full':
        return { allowed: true, requiresApproval: false, rule };

      default:
        return {
          allowed: false,
          requiresApproval: true,
          reason: 'Unknown policy',
        };
    }
  }

  /**
   * Request approval for a tool call
   */
  async requestApproval(context: ApprovalContext): Promise<ApprovalResult> {
    // Check immediate approval
    const immediate = this.checkApproval(context);
    if (immediate.allowed) {
      return immediate;
    }
    if (!immediate.requiresApproval) {
      return immediate;
    }

    // Create approval request
    const timeout = immediate.rule?.approvalTimeout ?? DEFAULT_APPROVAL_TIMEOUT;
    const request: ApprovalRequest = {
      id: `approval_${Date.now()}_${Math.random().toString(36).slice(2)}`,
      ruleId: immediate.rule?.id,
      toolName: context.toolName,
      toolArgs: context.toolArgs,
      sessionId: context.sessionId,
      senderId: context.senderId,
      channelId: context.channelId,
      agentId: context.agentId,
      status: 'pending',
      createdAt: new Date().toISOString(),
      expiresAt: new Date(Date.now() + timeout).toISOString(),
    };

    this.requests.set(request.id, request);

    // Wait for approval
    const approved = await this.waitForApproval(request.id, timeout);

    if (approved) {
      return {
        allowed: true,
        requiresApproval: true,
        rule: immediate.rule,
        requestId: request.id,
        reason: 'Approved by user',
      };
    }

    return {
      allowed: false,
      requiresApproval: true,
      rule: immediate.rule,
      requestId: request.id,
      reason: request.status === 'expired' ? 'Approval timed out' : 'Rejected by user',
    };
  }

  /**
   * Wait for approval response
   */
  private waitForApproval(requestId: string, timeout: number): Promise<boolean> {
    return new Promise((resolve) => {
      const timeoutId = setTimeout(() => {
        this.approvalCallbacks.delete(requestId);
        const request = this.requests.get(requestId);
        if (request && request.status === 'pending') {
          request.status = 'expired';
        }
        resolve(false);
      }, timeout);

      this.approvalCallbacks.set(requestId, { resolve, timeout: timeoutId });
    });
  }

  /**
   * Approve a pending request
   */
  approveRequest(requestId: string, approvedBy?: string): boolean {
    const request = this.requests.get(requestId);
    if (!request || request.status !== 'pending') {
      return false;
    }

    request.status = 'approved';
    request.resolvedAt = new Date().toISOString();
    request.resolvedBy = approvedBy;

    const callback = this.approvalCallbacks.get(requestId);
    if (callback) {
      clearTimeout(callback.timeout);
      callback.resolve(true);
      this.approvalCallbacks.delete(requestId);
    }

    return true;
  }

  /**
   * Reject a pending request
   */
  rejectRequest(requestId: string, reason?: string, rejectedBy?: string): boolean {
    const request = this.requests.get(requestId);
    if (!request || request.status !== 'pending') {
      return false;
    }

    request.status = 'rejected';
    request.resolvedAt = new Date().toISOString();
    request.resolvedBy = rejectedBy;
    request.reason = reason;

    const callback = this.approvalCallbacks.get(requestId);
    if (callback) {
      clearTimeout(callback.timeout);
      callback.resolve(false);
      this.approvalCallbacks.delete(requestId);
    }

    return true;
  }

  /**
   * Get pending approval requests
   */
  getPendingRequests(): ApprovalRequest[] {
    const now = new Date().toISOString();
    return Array.from(this.requests.values()).filter(
      (r) => r.status === 'pending' && r.expiresAt > now
    );
  }

  /**
   * Get an approval request
   */
  getRequest(requestId: string): ApprovalRequest | undefined {
    return this.requests.get(requestId);
  }

  /**
   * Clean up expired requests
   */
  cleanup(): number {
    const now = new Date();
    let count = 0;

    for (const [id, request] of this.requests) {
      if (request.status === 'pending' && new Date(request.expiresAt) < now) {
        request.status = 'expired';
        const callback = this.approvalCallbacks.get(id);
        if (callback) {
          clearTimeout(callback.timeout);
          callback.resolve(false);
          this.approvalCallbacks.delete(id);
        }
        count++;
      }
    }

    return count;
  }

  // Private methods

  private findMatchingRule(context: ApprovalContext): ApprovalRule | undefined {
    const rules = this.getRules().filter((r) => r.enabled);

    for (const rule of rules) {
      if (this.ruleMatches(context, rule)) {
        return rule;
      }
    }

    return undefined;
  }

  private ruleMatches(context: ApprovalContext, rule: ApprovalRule): boolean {
    // Check tool scope
    if (rule.tools && rule.tools.length > 0) {
      if (!rule.tools.includes(context.toolName)) {
        return false;
      }
    }

    // Check channel scope
    if (rule.channels && rule.channels.length > 0) {
      if (!context.channelId || !rule.channels.includes(context.channelId)) {
        return false;
      }
    }

    // Check sender scope
    if (rule.senders && rule.senders.length > 0) {
      if (!context.senderId || !rule.senders.includes(context.senderId)) {
        return false;
      }
    }

    // Check agent scope
    if (rule.agents && rule.agents.length > 0) {
      if (!context.agentId || !rule.agents.includes(context.agentId)) {
        return false;
      }
    }

    return true;
  }

  private isBlocked(context: ApprovalContext, rule?: ApprovalRule): boolean {
    if (!rule) return false;

    // Check blocked tools
    if (rule.blockedTools?.includes(context.toolName)) {
      return true;
    }

    // Check blocked commands (in args)
    if (rule.blockedCommands && context.toolArgs.command) {
      const cmd = String(context.toolArgs.command);
      if (rule.blockedCommands.some((bc) => cmd.includes(bc))) {
        return true;
      }
    }

    // Check blocked patterns
    if (rule.blockedPatterns) {
      const argsStr = JSON.stringify(context.toolArgs);
      if (rule.blockedPatterns.some((p) => new RegExp(p).test(argsStr))) {
        return true;
      }
    }

    return false;
  }

  private isAllowed(context: ApprovalContext, rule?: ApprovalRule): boolean {
    if (!rule) return false;

    // Check allowed tools
    if (rule.allowedTools?.includes(context.toolName)) {
      return true;
    }

    // Check allowed commands
    if (rule.allowedCommands && context.toolArgs.command) {
      const cmd = String(context.toolArgs.command);
      if (rule.allowedCommands.some((ac) => cmd.startsWith(ac))) {
        return true;
      }
    }

    // Check allowed patterns
    if (rule.allowedPatterns) {
      const argsStr = JSON.stringify(context.toolArgs);
      if (rule.allowedPatterns.some((p) => new RegExp(p).test(argsStr))) {
        return true;
      }
    }

    return false;
  }
}

export function createApprovalManager(): ApprovalManager {
  return new ApprovalManager();
}

File: repos/openrappter/python/openrappter/security/approvals.py
"""
ApprovalManager — execution policy and approval request management.

Policies are evaluated in priority order (highest first). The first matching
enabled rule governs the outcome. If no rule matches, the default_policy applies.
"""

import re
import time
import uuid
from dataclasses import dataclass, field
from typing import Dict, List, Literal, Optional

ApprovalPolicy = Literal['deny', 'allowlist', 'full']


@dataclass
class ApprovalRule:
    """A named policy rule with optional scope constraints."""
    id: str
    name: str
    policy: ApprovalPolicy
    allowed_tools: Optional[List[str]] = None
    tools: Optional[List[str]] = None          # scope: tool_name must be in this list
    channels: Optional[List[str]] = None       # scope: channel_id must be in this list
    agents: Optional[List[str]] = None         # scope: agent_id must be in this list
    blocked_patterns: Optional[List[str]] = None  # regex strings tested against str(tool_args)
    priority: int = 0
    enabled: bool = True


@dataclass
class ApprovalRequest:
    """A pending (or resolved) approval request for a tool invocation."""
    id: str
    tool_name: str
    tool_args: Dict
    channel_id: Optional[str]
    agent_id: Optional[str]
    status: Literal['pending', 'approved', 'rejected']
    reason: Optional[str]
    created_at: float


@dataclass
class ApprovalContext:
    """Contextual information about a tool call being evaluated."""
    tool_name: str
    tool_args: Dict = field(default_factory=dict)
    channel_id: Optional[str] = None
    agent_id: Optional[str] = None


class ApprovalManager:
    """
    Manages execution policies and approval workflows.

    Rules are matched in descending priority order. The first enabled rule
    whose scope conditions all match is applied. Scope conditions:
      - tools:    if present, tool_name must be in the list
      - channels: if present, channel_id must be in the list
      - agents:   if present, agent_id must be in the list

    Within a matching rule:
      - 'deny'      → always blocked
      - 'full'      → always allowed (unless a blocked_pattern fires)
      - 'allowlist' → allowed only if tool_name is in allowed_tools

    blocked_patterns are regex strings tested against str(tool_args). A match
    blocks the request regardless of policy.
    """

    def __init__(self, default_policy: ApprovalPolicy = 'deny') -> None:
        self._default_policy: ApprovalPolicy = default_policy
        self._rules: List[ApprovalRule] = []
        self._requests: Dict[str, ApprovalRequest] = {}

    # ------------------------------------------------------------------
    # Rule management
    # ------------------------------------------------------------------

    def add_rule(self, rule: ApprovalRule) -> None:
        """Append a rule to the manager."""
        self._rules.append(rule)

    def remove_rule(self, rule_id: str) -> None:
        """Remove the rule with the given id (no-op if not found)."""
        self._rules = [r for r in self._rules if r.id != rule_id]

    def get_rules(self) -> List[ApprovalRule]:
        """Return all registered rules."""
        return list(self._rules)

    # ------------------------------------------------------------------
    # Policy evaluation
    # ------------------------------------------------------------------

    def _rule_matches_scope(self, rule: ApprovalRule, ctx: ApprovalContext) -> bool:
        """Return True if all scope conditions on the rule are satisfied."""
        if rule.tools is not None and ctx.tool_name not in rule.tools:
            return False
        if rule.channels is not None and ctx.channel_id not in rule.channels:
            return False
        if rule.agents is not None and ctx.agent_id not in rule.agents:
            return False
        return True

    def _check_blocked_patterns(self, rule: ApprovalRule, ctx: ApprovalContext) -> Optional[str]:
        """
        Return a reason string if any blocked_pattern matches str(tool_args),
        otherwise None.
        """
        if not rule.blocked_patterns:
            return None
        args_str = str(ctx.tool_args)
        for pattern in rule.blocked_patterns:
            if re.search(pattern, args_str):
                return f"Blocked by pattern: {pattern}"
        return None

    def check_approval(self, context: ApprovalContext) -> Dict:
        """
        Evaluate policies against the given context.

        Returns a dict: { 'allowed': bool, 'reason': str }
        """
        # Sort by priority descending; stable sort preserves insertion order for ties
        sorted_rules = sorted(
            [r for r in self._rules if r.enabled],
            key=lambda r: r.priority,
            reverse=True,
        )

        for rule in sorted_rules:
            if not self._rule_matches_scope(rule, context):
                continue

            # Check blocked_patterns first — they override the policy
            blocked_reason = self._check_blocked_patterns(rule, context)
            if blocked_reason:
                return {'allowed': False, 'reason': blocked_reason}

            if rule.policy == 'deny':
                return {'allowed': False, 'reason': f"Denied by rule '{rule.name}'"}

            if rule.policy == 'full':
                return {'allowed': True, 'reason': f"Allowed by rule '{rule.name}'"}

            if rule.policy == 'allowlist':
                allowed_tools = rule.allowed_tools or []
                if context.tool_name in allowed_tools:
                    return {'allowed': True, 'reason': f"Allowed by rule '{rule.name}'"}
                return {
                    'allowed': False,
                    'reason': f"Tool '{context.tool_name}' not in allowlist for rule '{rule.name}'",
                }

        # No rule matched — apply default policy
        if self._default_policy == 'full':
            return {'allowed': True, 'reason': 'Allowed by default policy'}
        if self._default_policy == 'allowlist':
            return {'allowed': False, 'reason': 'Tool not in default allowlist'}
        return {'allowed': False, 'reason': 'Denied by default policy'}

    # ------------------------------------------------------------------
    # Approval request workflow
    # ------------------------------------------------------------------

    def request_approval(self, context: ApprovalContext) -> ApprovalRequest:
        """Create and store a new pending approval request."""
        request = ApprovalRequest(
            id=str(uuid.uuid4()),
            tool_name=context.tool_name,
            tool_args=context.tool_args,
            channel_id=context.channel_id,
            agent_id=context.agent_id,
            status='pending',
            reason=None,
            created_at=time.time(),
        )
        self._requests[request.id] = request
        return request

    def approve_request(self, request_id: str) -> None:
        """Mark the given request as approved."""
        request = self._requests.get(request_id)
        if request is not None:
            request.status = 'approved'

    def reject_request(self, request_id: str, reason: Optional[str] = None) -> None:
        """Mark the given request as rejected with an optional reason."""
        request = self._requests.get(request_id)
        if request is not None:
            request.status = 'rejected'
            request.reason = reason

    def get_pending_requests(self) -> List[ApprovalRequest]:
        """Return all requests with status='pending'."""
        return [r for r in self._requests.values() if r.status == 'pending']

    def get_request(self, request_id: str) -> Optional[ApprovalRequest]:
        """Return the request with the given id, or None."""
        return self._requests.get(request_id)
